### PR TITLE
bus/vcs: Use memory banks

### PR DESCRIPTION
**Pull request by wilbertpol**

### CODE REVIEW

File: src/devices/bus/vcs/rom.cpp [hunk 1]
```diff
@@ -624,44 +417,53 @@ void a26_rom_3f_device::write_bank(offs_t offset, uint8_t data)
...
+	space->install_rom(0x1c00, 0x1fff,get_rom_base() + 7 * 0x400);
+	space->install_write_handler(0x1fe0, 0x1fe7, write8sm_delegate(*this, FUNC(a26_rom_e0_device::switch_bank0)));
+	space->install_write_handler(0x1fe8, 0x1fef, write8sm_delegate(*this, FUNC(a26_rom_e0_device::switch_bank1)));
+	space->install_write_handler(0x1ff0, 0x1ff7, write8sm_delegate(*this, FUNC(a26_rom_e0_device::switch_bank2)));
+	space->install_read_tap(0x1fe0, 0x1fe7, "bank",
+			[this] (offs_t address, u8 &, u8) { if (!machine().side_effects_disabled()) switch_bank0(address, 0); });
+	space->install_read_tap(0x1fe8, 0x1fef, "bank",
+			[this] (offs_t address, u8 &, u8) { if (!machine().side_effects_disabled()) switch_bank1(address, 0); });
+	space->install_read_tap(0x1ff0, 0x1ff7, "bank",
+			[this] (offs_t address, u8 &, u8) { if (!machine().side_effects_disabled()) switch_bank2(address, 0); });
```
Comment (cuavas) [inline src/devices/bus/vcs/rom.cpp:451]: Do you actually need three trampolines for setting the banks?  You could use a `memory_bank_array_creator` for the banks and then use a single function with the bank number as a template parameter.

**Review by cuavas — COMMENTED** [review]
This is good cleanup.

File: src/devices/bus/vcs/harmony_melody.h [hunk 2]
```diff
@@ -31,6 +31,7 @@ class a26_rom_harmony_device : public a26_rom_f6_device
 	void harmony_arm7_map(address_map &map);
 
 	required_device<lpc210x_device> m_cpu;
+	uint8_t m_base_bank;
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/devices/bus/vcs/harmony_melody.h:35]: I totally missed that one. It's registered now.

File: src/devices/bus/vcs/scharger.cpp [hunk 3]
```diff
@@ -174,51 +153,69 @@ uint8_t a26_rom_ss_device::read(offs_t offset)
...
+uint8_t a26_rom_ss_device::read_cass(offs_t offset)
+{
+	// prevent spurious RAM write
+	m_address_bus_changes = RAM_WRITE_TIME + 1;
+
+	if (m_cassette->input() < 0)
+		return 0x00;
+	else
+		return 0x01;
+}
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/devices/bus/vcs/scharger.cpp]: Added newline

File: src/devices/bus/vcs/rom.cpp [hunk 4]
```diff
@@ -44,6 +44,22 @@ DEFINE_DEVICE_TYPE(A26_ROM_X07,   a26_rom_x07_device,   "vcs_x07",   "Atari VCS
 
 
 
+bool a26_rom_base_device::is_ram_present(size_t minimum_size)
+{
+	return !m_ram.empty() && m_ram.size() >= minimum_size;
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/devices/bus/vcs/rom.cpp]: Indeed, that is a bit confusing. I have removed the empty() check.

File: src/devices/bus/vcs/rom.cpp [hunk 5]
```diff
@@ -58,12 +74,10 @@ a26_rom_2k_4k_device::a26_rom_2k_4k_device(const machine_config &mconfig, const
...
-	space->install_read_handler(0x1000, 0x1fff, read8sm_delegate(*this, FUNC(a26_rom_2k_4k_device::read)));
-}
-
-uint8_t a26_rom_2k_4k_device::read(offs_t offset)
-{
-	return m_rom[offset & (m_rom_size - 1)];
+	if (m_rom_size < 0x1000)
+		space->install_rom(0x1000, 0x17ff, 0x800, get_rom_base());
+	else
+		space->install_rom(0x1000, 0x1fff, get_rom_base());
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/devices/bus/vcs/rom.cpp:80]: Only certain sizes are supported through vcs_slot.
For the 2k_4k type only 2KB and 4KB sized ROMs are allowed.

File: src/devices/bus/vcs/rom.cpp [hunk 6]
```diff
@@ -89,68 +103,32 @@ a26_rom_f6_device::a26_rom_f6_device(const machine_config &mconfig, const char *
...
-		}
-	}
-
-	return m_rom[offset + (m_base_bank * 0x1000)];
+	m_bank->configure_entries(0, 4, get_rom_base(), 0x1000);
+	space->install_read_bank(0x1000, 0x1fff, m_bank);
+	space->install_write_handler(0x1ff6, 0x1ff9, write8sm_delegate(*this, FUNC(a26_rom_f6_device::switch_bank)));
+	space->install_read_tap(0x1ff6, 0x1ff9, "bank",
+			[this] (offs_t address, u8 &, u8) { if (!machine().side_effects_disabled()) switch_bank(address - 0x1ff6, 0); });
+	install_super_chip_handlers(space);
```
[the remark this answers is not in the document]
Comment (wilbertpol) [reply src/devices/bus/vcs/rom.cpp]: I have swapped the install_super_chip_handler and the installation of the read tap in all devices that support the super chip.

File: src/devices/bus/vcs/rom.cpp (hunk 1 above)
Comment (wilbertpol) [reply src/devices/bus/vcs/rom.cpp]: Thanks for the pointer, changed to memory_bank_array_creator.

**Pull request merged**